### ISSUE TITLE
[Infra] Global country expansion — 195 countries, max data depth, full history

### ISSUE THREAD

**Issue by NathanLuckock** [Nathan7108/HackathonPractice#22]
## Problem
Sentinel monitors only 8 countries. A Palantir-grade platform monitors every country on earth. When a judge clicks any country on the map, it must return a real ML result.

## Goal
Expand from 8 countries to ALL ~195 countries. Max data depth for every source. Pre-cache everything before the hackathon.

### In scope
1. **data/countries.json** — master reference for all ~195 countries (ISO2, ISO3, name, lat/lng centroid, region, acled_name)
2. **GDELT** — download full v2 history (2015-present), single-pass extraction for ALL countries
3. **ACLED** — fetch from 1997-present for every country ACLED covers
4. **UCDP** — split existing global GED + armed conflict CSVs into per-country files
5. **World Bank** — fetch mrv=30 (30 years) for all ~195 countries
6. **pipeline.py** — MONITORED_COUNTRIES loads from countries.json dynamically
7. **Anomaly detectors** — retrain for every country with sufficient GDELT data
8. **XGBoost** — retrain on expanded multi-country dataset
9. **fetch_all_countries.py** — master orchestration script that runs everything

### Out of scope
- LSTM retraining (existing model generalizes)
- Frontend changes (separate sprint)

## Acceptance Criteria
- [ ] data/countries.json has ~195 entries with lat/lng centroids
- [ ] GDELT: data/gdelt/{CC}_events.csv for every country with events (expect 150+)
- [ ] ACLED: data/acled/{country}.csv for every country with conflict data
- [ ] UCDP: data/ucdp/{country}_ged.csv split from master dataset
- [ ] World Bank: data/world_bank/{ISO3}.json for all ~195 countries
- [ ] MONITORED_COUNTRIES dynamically loaded from countries.json
- [ ] Anomaly models trained for all countries with >= 5 weeks GDELT data
- [ ] XGBoost retrained on all countries, accuracy > 80%
- [ ] python -m backend.ml.data.fetch_all_countries runs end-to-end
- [ ] SentinelFeaturePipeline.compute_all_countries() returns features for all countries

### Step 1: data/countries.json

Generate from pycountry + manual lat/lng centroids. Include every UN member state plus Taiwan, Kosovo, Palestine.

```python
# scripts/generate_countries.py
import pycountry
import json

# Country centroids (approximate) — use a static dataset or hardcode
# Many sources: https://developers.google.com/public-data/docs/canonical/countries_csv
# Or pip install pycountry + manual lat/lng

CENTROIDS = {
    "AF": (33.93, 67.71), "AL": (41.15, 20.17), "DZ": (28.03, 1.66),
    # ... full list needed
}

# ACLED name overrides (where ACLED uses different names)
ACLED_OVERRIDES = {
    "US": "United States of America",
    "GB": "United Kingdom",
    "RU": "Russia",
    "KR": "Republic of Korea",
    "KP": "Democratic People's Republic of Korea",
    "IR": "Iran",
    "SY": "Syria",
    "VE": "Venezuela",
    "BO": "Bolivia",
    "TZ": "Tanzania",
    "CD": "Democratic Republic of Congo",
    "CG": "Republic of Congo",
    "CI": "Ivory Coast",
    "LA": "Laos",
    "MM": "Myanmar",
    "TW": "Taiwan",  # Not in pycountry
    "XK": "Kosovo",  # Not in pycountry
    "PS": "Palestine", # Not in pycountry
}

countries = []
for c in pycountry.countries:
    iso2 = c.alpha_2
    lat, lng = CENTROIDS.get(iso2, (0, 0))
    countries.append({
        "iso2": iso2,
        "name": c.name,
        "iso3": c.alpha_3,
        "lat": lat,
        "lng": lng,
        "region": "",  # Can fill from pycountry subdivision or manual
        "acled_name": ACLED_OVERRIDES.get(iso2, c.name),
    })

# Add Taiwan, Kosovo, Palestine manually
manual = [
    {"iso2": "TW", "name": "Taiwan", "iso3": "TWN", "lat": 23.70, "lng": 120.96, "region": "Asia", "acled_name": "Taiwan"},
    {"iso2": "XK", "name": "Kosovo", "iso3": "XKX", "lat": 42.60, "lng": 20.90, "region": "Europe", "acled_name": "Kosovo"},
    {"iso2": "PS", "name": "Palestine", "iso3": "PSE", "lat": 31.95, "lng": 35.23, "region": "Middle East", "acled_name": "Palestine"},
]
countries.extend(manual)

with open("data/countries.json", "w") as f:
    json.dump(countries, f, indent=2)
print(f"Generated {len(countries)} countries")
```

For lat/lng centroids, use Google's public dataset:
https://developers.google.com/public-data/docs/canonical/countries_csv
Download it and merge.

### Step 2: GDELT — full history, single-pass global extraction

Rewrite fetch_gdelt.py to add a `fetch_gdelt_all_countries()` function:

```python
from collections import defaultdict

def fetch_gdelt_all_countries(days_back=3650):
    """
    Download GDELT v2 zips ONCE, extract events for ALL countries in single pass.
    days_back=3650 = ~10 years (full GDELT v2 history from 2015).
    
    This is dramatically faster than per-country fetching because each zip
    is downloaded and parsed once, then events are partitioned by country code.
    """
    urls = _get_export_urls(days_back)
    data_dir = _data_dir()
    
    # Process in chunks to manage memory
    CHUNK_SIZE = 500  # Process 500 zips at a time
    
    # Track which countries we've seen
    country_files = {}  # cc -> file handle for appending
    
    for chunk_start in range(0, len(urls), CHUNK_SIZE):
        chunk_urls = urls[chunk_start:chunk_start + CHUNK_SIZE]
        country_frames = defaultdict(list)
        
        for url in tqdm(chunk_urls, desc=f"GDELT chunk {chunk_start//CHUNK_SIZE + 1}"):
            path = _download_or_get_cached(url)
            if path is None:
                continue
            try:
                with zipfile.ZipFile(path) as z:
                    with z.open(z.namelist()[0]) as f:
                        df = pd.read_csv(
                            io.BytesIO(f.read()), sep="\t", header=None,
                            usecols=GDELT_COL_INDICES, names=GDELT_COL_NAMES,
                            dtype=str, on_bad_lines="skip", low_memory=False,
                        )
                # Partition by country codes found in Actor1 and Actor2
                a1 = df["Actor1CountryCode"].fillna("").str.upper()
                a2 = df["Actor2CountryCode"].fillna("").str.upper()
                all_codes = set(a1.unique()) | set(a2.unique())
                all_codes.discard("")
                
                for cc in all_codes:
                    mask = (a1 == cc) | (a2 == cc)
                    country_frames[cc].append(df[mask].copy())
            except Exception:
                continue
        
        # Write chunk results to CSVs (append mode)
        for cc, frames in country_frames.items():
            merged = pd.concat(frames, ignore_index=True)
            for col in ["GoldsteinScale", "NumMentions", "AvgTone"]:
                merged[col] = pd.to_numeric(merged[col], errors="coerce")
            out_path = data_dir / f"{cc}_events.csv"
            if out_path.exists():
                merged.to_csv(out_path, mode="a", header=False, index=False)
            else:
                merged.to_csv(out_path, index=False)
        
        print(f"  Chunk {chunk_start//CHUNK_SIZE + 1}: {len(country_frames)} countries")
    
    # Count results
    csvs = list(data_dir.glob("*_events.csv"))
    print(f"GDELT complete: {len(csvs)} country files")
```

**Key optimization:** Each zip is downloaded once, opened once, and ALL country codes are extracted simultaneously. The zips are already cached from previous fetches.

### Step 3: ACLED — full history, all countries

```python
def fetch_acled_all_countries(token: str, start_date: str = "1997-01-01"):
    """Fetch ACLED for every country in countries.json."""
    with open("data/countries.json") as f:
        countries = json.load(f)
    
    data_dir = Path("data/acled")
    data_dir.mkdir(parents=True, exist_ok=True)
    
    for entry in tqdm(countries, desc="ACLED"):
        acled_name = entry["acled_name"]
        safe = acled_name.lower().replace(" ", "_")
        out_path = data_dir / f"{safe}.csv"
        
        # Skip if already fetched
        if out_path.exists() and out_path.stat().st_size > 100:
            continue
        
        try:
            df = fetch_acled(acled_name, token, start_date=start_date)
            if len(df) > 0:
                df.to_csv(out_path, index=False)
                print(f"  {acled_name}: {len(df)} events")
            time.sleep(0.5)
        except Exception as e:
            print(f"  {acled_name}: {e}")
```

### Step 4: UCDP — split master into per-country files

```python
def split_ucdp_global():
    """Split UCDP GED master CSV into per-country files."""
    ucdp_dir = Path("data/ucdp")
    
    # Find the master GED file (could be various names)
    ged_files = list(ucdp_dir.glob("*ged*.csv")) + list(ucdp_dir.glob("*GED*.csv"))
    if not ged_files:
        print("No UCDP GED master file found")
        return
    
    # Use the largest file as master
    master_path = max(ged_files, key=lambda p: p.stat().st_size)
    print(f"Splitting {master_path.name}...")
    
    df = pd.read_csv(master_path, low_memory=False)
    country_col = "country" if "country" in df.columns else "country_name"
    
    for country, group in df.groupby(country_col):
        safe = country.lower().replace(" ", "_").replace("(", "").replace(")", "")
        out_path = ucdp_dir / f"{safe}_ged.csv"
        group.to_csv(out_path, index=False)
    
    print(f"Split into {df[country_col].nunique()} country files")
```

### Step 5: World Bank — all countries, 30 years

Update fetch_world_bank.py:
```python
def fetch_world_bank_all_countries(mrv=30):
    """Fetch 30 years of World Bank indicators for all countries."""
    with open("data/countries.json") as f:
        countries = json.load(f)
    
    wb_dir = Path("data/world_bank")
    wb_dir.mkdir(parents=True, exist_ok=True)
    
    for entry in tqdm(countries, desc="World Bank"):
        iso3 = entry["iso3"]
        out_path = wb_dir / f"{iso3}.json"
        
        if out_path.exists():
            continue
        
        try:
            features = fetch_world_bank_features(iso3)  # Update this to use mrv=30
            with open(out_path, "w") as f:
                json.dump({"iso3": iso3, "features": features}, f)
        except Exception as e:
            print(f"  {iso3}: {e}")
```

### Step 6: Pipeline — dynamic MONITORED_COUNTRIES

In pipeline.py, replace hardcoded dict:
```python
import json

def _load_countries() -> dict:
    path = Path(__file__).resolve().parents[2] / "data" / "countries.json"
    if path.exists():
        with open(path) as f:
            entries = json.load(f)
        return {
            e["iso2"]: {
                "name": e["name"],
                "iso3": e["iso3"],
                "acled_name": e.get("acled_name", e["name"]),
                "lat": e.get("lat", 0),
                "lng": e.get("lng", 0),
                "region": e.get("region", ""),
            }
            for e in entries
        }
    # Fallback
    return {"UA": {"name": "Ukraine", "iso3": "UKR", "acled_name": "Ukraine"}, ...}

MONITORED_COUNTRIES = _load_countries()
```

### Step 7: Retrain anomaly detectors for all countries

```python
def train_all_anomaly_detectors():
    """Train for every country with >= 5 weeks of GDELT data."""
    data_dir = Path("data/gdelt")
    count = 0
    for csv_path in sorted(data_dir.glob("*_events.csv")):
        cc = csv_path.stem.replace("_events", "")
        try:
            df = pd.read_csv(csv_path)
            df['date'] = pd.to_datetime(df['SQLDATE'].astype(str), format='%Y%m%d', errors='coerce')
            weeks = df['date'].dt.to_period('W').nunique()
            if weeks >= 5:
                train_anomaly_detector(cc, df)
                count += 1
        except Exception as e:
            print(f"  {cc}: {e}")
    print(f"Trained anomaly detectors for {count} countries")
```

### Step 8: Retrain XGBoost on expanded dataset

In risk_scorer.py, build_training_dataset() already loops MONITORED_COUNTRIES. Once pipeline.py loads all countries from countries.json, the training set automatically expands to include every country with ACLED data.

### Master orchestration script: backend/ml/data/fetch_all_countries.py

```python
"""
Master data fetcher — downloads and caches ALL data for ALL countries.
Run once before the hackathon. Takes a few hours total.

Usage:
    python -m backend.ml.data.fetch_all_countries
    python -m backend.ml.data.fetch_all_countries --step gdelt
    python -m backend.ml.data.fetch_all_countries --step acled
    python -m backend.ml.data.fetch_all_countries --step ucdp
    python -m backend.ml.data.fetch_all_countries --step worldbank
    python -m backend.ml.data.fetch_all_countries --step models
"""
if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument("--step", choices=["all", "countries", "gdelt", "acled", "ucdp", "worldbank", "models"], default="all")
    args = parser.parse_args()
    
    if args.step in ("all", "countries"):
        generate_countries_json()
    if args.step in ("all", "gdelt"):
        fetch_gdelt_all_countries(days_back=3650)  # Full GDELT v2 history
    if args.step in ("all", "acled"):
        token = get_acled_token(...)
        fetch_acled_all_countries(token, start_date="1997-01-01")  # Full ACLED history
    if args.step in ("all", "ucdp"):
        split_ucdp_global()
    if args.step in ("all", "worldbank"):
        fetch_world_bank_all_countries(mrv=30)  # 30 years
    if args.step in ("all", "models"):
        train_all_anomaly_detectors()
        # XGBoost retraining
        from backend.ml.risk_scorer import build_training_dataset, train_risk_scorer
        df = build_training_dataset()
        train_risk_scorer(df)
```

### Edge cases:
- ACLED name mismatches: some countries use different names. The ACLED_OVERRIDES dict handles this.
- Countries with zero events (e.g., Liechtenstein, Monaco): pipeline returns all-zero features → LOW risk. Correct behavior.
- GDELT memory: processing all zips at once could use a lot of RAM. Process in chunks of 500 zips.
- Disk space: GDELT full history CSVs for all countries: ~5-10GB. ACLED: ~2GB. World Bank: <100MB. UCDP: <500MB.
- Rate limiting: ACLED 0.5s between requests. World Bank no limit. GDELT no limit (direct download).

## Manual Test Steps
1. python -m backend.ml.data.fetch_all_countries --step countries → verify data/countries.json has ~195 entries
2. python -m backend.ml.data.fetch_all_countries --step gdelt → verify data/gdelt/ has 150+ CSV files
3. python -m backend.ml.data.fetch_all_countries --step acled → verify data/acled/ has CSVs
4. python -m backend.ml.data.fetch_all_countries --step ucdp → verify data/ucdp/ has per-country files
5. python -m backend.ml.data.fetch_all_countries --step worldbank → verify data/world_bank/ has ~195 JSONs
6. python -m backend.ml.data.fetch_all_countries --step models → verify models/ has anomaly + retrained risk_scorer
7. python -c "from backend.ml.pipeline import MONITORED_COUNTRIES; print(len(MONITORED_COUNTRIES))" → ~195

---

## Cursor Agent Instructions

Repo: Nathan7108/HackathonPractice
Working directory: C:\Users\natha\HackathonPractice
Branch: feature/global-country-expansion

### STEP 0 — Read context
```
cat PROGRESS.md
cat backend/ml/pipeline.py
cat backend/ml/data/fetch_gdelt.py
cat backend/ml/data/fetch_acled.py
cat backend/ml/data/fetch_ucdp.py
cat backend/ml/data/fetch_world_bank.py
cat backend/ml/anomaly.py
cat backend/ml/risk_scorer.py
```

### STEP 1 — Create branch
```
git checkout main && git pull origin main
git checkout -b feature/global-country-expansion
```

### STEP 2 — Create data/countries.json
Generate via pycountry + Google centroids CSV. Include ~195 countries + Taiwan/Kosovo/Palestine. Include ACLED name overrides.

### STEP 3 — Update backend/ml/data/fetch_gdelt.py
Add fetch_gdelt_all_countries(days_back=3650). Single-pass extraction. Chunk processing. Append mode for CSVs.

### STEP 4 — Update backend/ml/data/fetch_acled.py
Add fetch_acled_all_countries(token, start_date="1997-01-01"). Skip already-fetched. 0.5s delay.

### STEP 5 — Update backend/ml/data/fetch_ucdp.py
Add split_ucdp_global(). Split master GED CSV by country column.

### STEP 6 — Update backend/ml/data/fetch_world_bank.py
Add fetch_world_bank_all_countries(mrv=30). Update fetch_world_bank_features() to accept mrv parameter.

### STEP 7 — Update backend/ml/pipeline.py
Replace hardcoded MONITORED_COUNTRIES with _load_countries() from data/countries.json. Keep fallback to original 8.

### STEP 8 — Update backend/ml/anomaly.py
Update train_all_anomaly_detectors() to iterate all GDELT CSVs in data/gdelt/, not just 8 hardcoded countries.

### STEP 9 — Create backend/ml/data/fetch_all_countries.py
Master orchestration with --step argument for running individual or all steps.

### STEP 10 — Validate
```
python -m backend.ml.data.fetch_all_countries --step countries
python -c "from backend.ml.pipeline import MONITORED_COUNTRIES; print(len(MONITORED_COUNTRIES))"
```
Verify ~195 countries loaded.

### STEP 11 — Commit and push
```
git add data/countries.json backend/ml/data/ backend/ml/pipeline.py backend/ml/anomaly.py
git commit -m "feat: expand to all ~195 countries — global data fetching + pipeline"
git push -u origin feature/global-country-expansion
```

### STEP 12 — Open PR
```
gh pr create --title "[Infra] Global country expansion — 195 countries" --body "Closes #22" --base main
```

### Definition of Done
- [ ] countries.json exists with ~195 entries
- [ ] fetch_all_countries.py --step countries works
- [ ] MONITORED_COUNTRIES loads dynamically
- [ ] fetch_gdelt_all_countries() function exists (actual download run separately)
- [ ] fetch_acled_all_countries() function exists
- [ ] split_ucdp_global() function exists
- [ ] fetch_world_bank_all_countries() function exists
- [ ] train_all_anomaly_detectors() iterates all GDELT CSVs
- [ ] PR created

### IMPORTANT: Do NOT run the full data downloads in this PR
The code changes and countries.json should be committed. The actual data downloads (GDELT, ACLED, etc.) are run separately by Nathan after merge because they take time and produce large files that aren't committed to git.

### If You Get Stuck
- pycountry: pip install pycountry
- ACLED names: check https://acleddata.com/data/ for exact country names
- GDELT country codes: 2-letter FIPS codes (mostly same as ISO2, some differ)
- World Bank ISO3 codes: same as pycountry alpha_3
- If memory issues with GDELT: reduce CHUNK_SIZE from 500 to 200
